pinot-core/src/main/java/org/apache/pinot/core/common/DataFetcher.java

```diff
@@ -309,13 +309,15 @@ private class ColumnValueReader implements AutoCloseable {
     final Dictionary _dictionary;
     final DataType _storedType;
     final boolean _singleValue;
+    final boolean _useDictionary;
 
     boolean _readerContextCreated;
     ForwardIndexReaderContext _readerContext;
 
     ColumnValueReader(ForwardIndexReader reader, @Nullable Dictionary dictionary) {
       _reader = reader;
       _dictionary = dictionary;
+      _useDictionary = _reader.isDictionaryEncoded() && _dictionary != null;
```

Comment:
(minor) Suggest putting this check on the caller side. We can also consider splitting it into 2 readers, one for dictionary encoded and one for raw